src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs

```diff
@@ -203,38 +266,131 @@ private void addMolecules(IMoleculeDependentBuilder builder, MoleculeList molecu
+         /// </summary>
+         public bool RequiresClone => BuildersToMerge.Count > 0;
+
+         /// <summary>
+         ///    It is required to also clone each builder being merged if we have 2 or more builders to merge.
+         ///    This prevents cross-contamination between builders during sequential merges.
+         ///    Example: When merging B1, then B2 into finalBuilder:
+         ///    - Merge B1: B1's children/references are transferred to finalBuilder
+         ///    - Merge B2: B2's merge operations might modify children that came from B1
+         ///    - If B1 wasn't cloned, these modifications propagate back to the original building block through shared references
+         ///    With only 1 builder to merge, there are no subsequent merges to cause side effects, so cloning is not needed.
+         /// </summary>
+         public bool RequiresMergeClone => BuildersToMerge.Count >= 2;
+
+         public BuilderMergeInfo(BuilderSource<T> baseBuilder) : this(baseBuilder, Enumerable.Empty<BuilderSource<T>>())
+         {
+         }
 
-      private IReadOnlyList<(T Builder, IBuildingBlock BuildingBlock)> allBuilderSources<T>(Func<Module, IBuildingBlock<T>> propAccess) where T : IBuilder =>
-         _simulationConfiguration.ModuleConfigurations
-            .Select(x => propAccess(x.Module))
-            .Where(x => x != null)
-            .SelectMany(x => x.Select(builder => (builder, (IBuildingBlock) x)))
+         public BuilderMergeInfo(BuilderSource<T> baseBuilder, IEnumerable<BuilderSource<T>> buildersToMerge)
+         {
+            BaseBuilder = baseBuilder;
+            BuildersToMerge = buildersToMerge.ToList();
+         }
+      }
+
+      private IReadOnlyList<BuilderMergeInfo<T>> analyzeBuilderMerges<T>(Func<Module, IBuildingBlock<T>> propAccess) where T : IBuilder
+      {
+         var buildingBlocksAndMergeBehaviors = buildingBlockAndMergeBehaviors(propAccess);
+
+         if (buildingBlocksAndMergeBehaviors.Count == 0)
+            return new List<BuilderMergeInfo<T>>();
+
+         var allBuildersWithBehaviors = buildingBlocksAndMergeBehaviors
+            .SelectMany(x => x.buildingBlock.Select(builder => (builder, x.buildingBlock, x.mergeBehavior)))
             .ToList();
 
+         var results = new List<BuilderMergeInfo<T>>();
+
+         foreach (var group in allBuildersWithBehaviors.GroupBy(x => x.builder.Name))
+         {
+            var builders = group.ToList();
+
+            //only one, we use it as is
+            if (builders.Count == 1)
+            {
+               var builderSource = new BuilderSource<T>(builders[0].builder, builders[0].buildingBlock);
+               results.Add(new BuilderMergeInfo<T>(builderSource));
+               continue;
+            }
+
+            //last one is overwrite. We will use it as is also
+            var (lastBuilder, lastBuildingBlock, mergeBehavior) = builders[builders.Count - 1];
+            if (mergeBehavior == MergeBehavior.Overwrite)
```

Comment:
@PavelBal @rwmcintosh @Yuri05 Correct me if I am wrong but I believe that we said that in case of overwrite, we use as is? correct? I know I was disagreeing with some suggestions but I can't find trace of it in the discussion. 
Extend will defacto combine. But one overwrite and it becomes the new source of truth